### ISSUE TITLE
Develop example of window splitter design pattern

### ISSUE THREAD

**Issue by mcking65** [w3c/aria-practices#130]
The window splitter design pattern is at:
http://w3c.github.io/aria-practices/#windowsplitter

Note that this issue depends on issue #129.

**Comment by lifeiscontent** [comment]
@mcking65 any updates?

**Comment by MichielBijl** [comment]
Given the lack of updates and my need for this—I’ve self-assigned this one.

**Comment by MichielBijl** [comment]
I’m struggling with drag support. Anyone that has experience with this up for a collab?

**Comment by lifeiscontent** [comment]
@MichielBijl I can help, want to send me an invite to a github repo?

**Comment by MrGrigri** [comment]
I've got a working drag example on [CodePen](https://codepen.io/MrGrigri/pen/vvEmYP). I was working on the drag part first and then I was going to add the keyboard interaction. I'm using CSS variables to control the width of the sidebar.

**Comment by ZoeBijl** [comment]
@lifeiscontent do you have a CodePen account? I can invite you to my example there. If not, I can put it in a repo.

@MrGrigri excellent! I’ll have a look at that, looks to be working well!

**Comment by MrGrigri** [comment]
Thanks @ZoeBijl. My example utilizes CSS Variables...so I'm not sure what the minimum requirements for W3C examples are.

**Comment by lifeiscontent** [comment]
@ZoeBijl I have an account, my username is the same as GitHub

**Comment by jjrice2233** [comment]
yes

**Comment by captainKAIO** [comment]
If I can help, I have done a complete example on codepen
https://codepen.io/captainKAIO/pen/yLYPPxX. 
A note, I had a hard time defining what the valuemax was.

**Comment by lifeiscontent** [comment]
This isn’t complete. It’s completely broken on mobile.

**Comment by captainKAIO** [comment]
My intention was to implement the pattern as described. The management of "touch" or "drag" events is not one of them, but I will integrate them, it's a good idea. On the other hand, responsive design would like the splitter to reposition itself under the element. Therefore, the orientation changes. Wouldn't this change of arrow keys be disturbing for a blind person?

**Comment by lifeiscontent** [comment]
@captainKAIO I don't think it makes sense to make it change in a responsive layout.

In the case of a window splitter, you have fixed viewport panes, and if the layout was to change based off screen size I think it would be confusing from both a technical perspective (developer would find it unreasonable) and also from a UX perspective.

**Comment by captainKAIO** [comment]
It's done. I didn't manage the swipe, but it works for a finger movement.

**Comment by lifeiscontent** [comment]
@captainKAIO now the mouse interactions are broken.

**Comment by captainKAIO** [comment]
It's a little bug, It's fixed now.

**Comment by captainKAIO** [comment]
Finally, I implemented the swipe gesture like the "end" and "home" keys, it's funny

**Comment by lifeiscontent** [comment]
@captainKAIO Nice job! The UX is perfect. The only other note I'd suggest is having the vertical splitter not be an H1, it should be a splitter itself no?

**Comment by captainKAIO** [comment]
Ok, I added a "pane" paragraph to explain the screen disposition.
The window splitter pattern needs two panes for the separator and I think H1 couldn't have been a splitter itself.

**Comment by captainKAIO** [comment]
I have a question to finalize my example with the F6 key. Is F6 used to browse all the panes present in the page or only the two associated panes?

**Comment by StommePoes** [comment]
You'll need to test which browsers already use F6, because I recall some do to move around browser chrome (address bar etc). For example Chrome on Windows uses it, and I have no idea what Appley things do.

I recall Slack had to pair the F6 key with something else, or it was the reason why they also have CTRL+grave-accent, for a more cross-browser keystroke pair.

**Comment by carmacleod** [comment]
Firefox also uses F6, and if you type Ctrl+H to open the History side panel, you will see that F6 cycles between the Firefox URL bar, the document body, and the History panel. Last time I checked, Firefox also cycled through old (deprecated!) `<frame>` elements if there were any on the page. I don't recall if they also included `<iframe>` elements in the cycle, or maybe that was Chrome (or both).

Anyhow, I'm not sure whether the APG advice to use F6 for cycling between splitter panes is final.
So, given that it is optional, it might be best to leave it out for now.

However, I would definitely recommend the following for your particular example:
- change the `<div id="pane1">` (i.e. the "Table of contents") to `<nav id="pane1">`
- change the `<div id="pane2">` to `<main id="pane2">`
- maybe change the `<div id="primary">` to `<header id="primary">` ? (maybe with the h1 inside the header?)

That way, the panes will all be _landmarks_, and screen reader users will be able to navigate to any one of them using their screen reader's landmark navigation commands. Keyboard users who are not using a screen reader can navigate to landmarks using the [Landmarks plugin](http://matatk.agrip.org.uk/landmarks/), and someday browsers will figure out that landmark navigation is _useful_, and they will provide a native keyboard shortcut. ;)

**Comment by StommePoes** [comment]
Can I respectfully disagree?
As a sighted and mostly-keyboard user, who has had the sad experience of navigating dashboards... as much as I hate (in-browser) Slack's unfriendliness as far as being able to see things and scroll things at my zoom level (okay, actually, I can't. Slack's unusable if I don't have access to a yuuuge monitor), it is absolutely a godsend to be able to move between panes in whatever browser I'm in, even someone else's computer or a public (library) machine.

I feel so long as developers continue to insist on recreating desktop application windows on the web, they're required to make them work with keyboard in an approximate amount of efficiency as desktop application windows. I feel panes/panels should have a conventional or nearly-conventional set of keystrokes that allow focus to be moved between them and not rely on all keyboarders being screen reader users or happening to have administrative privileges to add plugins or other modifications to software to avoid the Thousand Tabs of Death.

I mean vimperator was nice but I can't imagine telling students in a school they need to stick to a particular browser so they can use a website. We already have that problem regarding AT-browser compatibility :(

**Comment by sinabahram** [comment]
I must agree with @StommePoes  on this issue. I also do not feel it is appropriate to ignore a pattern that has existed for decades, which helps keyboard users save so very many keystrokes. F6 is a wonderful affordance that facilitates this. While it is true that Apple is making their laptops less usable and less accessible to keyboard users by making function keys far harder to access, I hope we don't fall into a pattern where the suboptimal (in my opinion) usability decisions of one manufacturer adversely affects the entire web for all. I bring that up as it is the usual objection to including f6 advice.

**Comment by carmacleod** [comment]
@StommePoes 
> Can I respectfully disagree?

Absolutely! And thank-you for making me go take a closer look at the Slack markup... I thought that their "F6 desktop/Ctrl+F6 web" behavior was navigating strictly through landmarks (that mostly happen to be in panes). Now that I look more closely, I see that even though 3 of their "panes" are indeed marked up as `<nav>`, `<div role="navigation">`, and `<div role="main">`, they are "cheating" in a couple of other places and just jumping to div-things that are not properly marked up as landmark regions. I will grump at them for that. They need to make those be `<section>` or `<div role="region">` and give them an accessible label so that screen reader users can use their landmark navigation keys to get to them.

We actually agree with each other, you and I, but I have an ulterior motive. ;)
Sorry I didn't explain it earlier.

Here's the back story: I would like the browsers to support keyboard landmark navigation. All of them. Natively. For everybody.
I will be working more on this in the near future, but for now, here's a github issue that has a bit of a summary (feel free to give it a thumbs-up 😄): https://github.com/WebWeWant/webwewant.fyi/issues/63

In the web-a11y slack, there's also this post: https://web-a11y.slack.com/archives/C042TSFGN/p1596129362456100

The most likely keystroke to use for landmark navigation on the web would be Ctrl+F6 (and Ctrl+Shift+F6 for reverse). Browsers have already taken F6/Shift+F6... and even though the behavior is similar-ish, it probably doesn't make sense to overload it. Also, between the platforms and the browsers, there's not a lot of other possible keystrokes. Maybe some combination of backtick (grave-accent), as you mentioned. Note that Slack's Alt+backtick goes directly to the message text input, and does not navigate through the landmarks.

So, back to the Window Splitter pattern discussion in this issue, and my ulterior motive. Since F6 is pretty much taken by browsers, I was worried that Ctrl+F6 might come up in this conversation, and I very much want to keep that one available as a possible keystroke for native landmark navigation. Mind you, if it's possible to use F6 (and Shift+F6) to successfully navigate through the browser trim and then through the web content (panes, frames, and maybe iframes) and then back through browser trim, and if that doesn't feel weird, then by all means use F6 (without the Ctrl modifier) to do so. Come to think of it, it would be a good experiment to try it with the current example, as long as it's tested in Chrome, Edge, FF, and Safari. It would be great to know whether or not that works out.

On a related note, looking at the [APG advice for Window Splitter panes](https://w3c.github.io/aria-practices/#windowsplitter), I see that there's no semantic markup suggested for the panes themselves (other than they happen to be elements that have a splitter between them), and I think we should be giving advice for panes to be semantically marked up as landmarks, because that's really what they are.

**Comment by carmacleod** [comment]
@sinabahram I noticed your comment as I was finishing up my response to @StommePoes - hopefully I have explained my thinking well enough, above.

However, now I have a follow-on question for both of you:
Do you actually use F6 on the web? If so, where? (other than just "go to URL bar", "go to body")?

**Comment by sinabahram** [comment]
@carmacleod  that does help a lot, but I think we should be careful, as you stated, not to conflate f6 and ctrl+f6 with your absolutely wonderful mission to promote landmark navigation. I feel like those can be thought of as similar goals but not until recently would people maybe map them as such. For example, you saw that with Slack. I've flagged those unlabeled regions to them several times to no avail, by the way, so good luck to you friend and thanks for reporting them (more voices make a difference). 

In addition to Slack, we used f6 really affectively to make the various areas of the WeScheme blocks project accessible. That's the treeview example with the online IDE whose link I sent into the Zoom on our previous to last ARIA deep dive call, "Trees, part 2". In that context, the students could use f6 to navigate between the definitions area, code area, and error messages from the compiler, which totally freed us up to use tab, shift+tab, and other keystrokes to facilitate better accessibility within their respective sections. Plus, f6 was necessary as tab is super-tricky when we talk about editors that must implement indent support like in a REPL where you want to use tab for autocompletion and then shift+f6 into the console output, then f6 back to the editing area.

On desktop applications, it's obviously widely used, and as @StommePoes  points out, devs are replicating desktop experiences on the web, and have been for years, so I want to at least have the same mediocre  level of accessibility we all have on desktop, not something worse on the web *sad but hopeful smile*. Hope that helps.

**Comment by carmacleod** [comment]
@sinabahram So, with WeScheme, were you able to take over the F6 key without anyone getting upset with you for stealing their favorite "go to URL bar" shortcut? Or did you somehow (maybe by selectively calling event.preventDefault() in the F6 key event handler??) manage to incorporate the browser's F6 navigation ring into the WeScheme panes navigation ring?
If you stole the key, then were you able to tell people "Sorry, we stole the F6 key for our app, so you'll just have to get used to using Ctrl+L/Cmd+L to go to the browser's URL bar"? I'm just not sure that's going to fly if all web apps start doing that. People get used to their shortcut keys. :)

However, your point is taken that we should be careful not to conflate f6 and ctrl+f6 with the mission to promote landmark navigation. :)

We should run a contest for the best keystroke suggestion for landmark navigation. Has to be something that's not already used by the platform, or the browsers, or any screen readers or other AT. Need to be able to pair it with Shift key for reverse navigation. Likely can't use Alt key (without stealing), because Mac already stole most (all??) of the Alt key combinations for entering special characters into text fields, and we want to be able to navigate out of those. Ideally shouldn't use function keys because of recent Mac keyboards only having virtual function keys. Should be ergonomically feasible on at least a majority of keyboards of all languages. If anyone out there knows of such a key shortcut, please pass along the knowledge!

I'm only half-kidding about the contest. Winner gets recognition and lots of appreciation and they get to proudly say, "That's _my_ key shortcut!". ;)

**Comment by sinabahram** [comment]
@carmacleod, we were targeting Chrome rather heavily for that project, just due to Chromebooks and other reasons, but I remember checking that there were at least three other ways to get to the address bar e.g. alt+d, ctrl+l, and even with shift+tab from top of document. I don't know if I have a good sense of how many folks use f6 to get to the address bar as opposed to those methods instead, but wen three other ways existed, it felt like an acceptable trade off.

As for not using function keys because of Mac, I'm really troubled by this. I'm deeply frustrated by a similar notion that came from our treeview conversation that we must use the same keys across platforms. That one regarded enter and space. I find this a losing proposition doomed to absolute mediocrity in the best case, and it just goes downhill from there.

My Frustration:
Apple removes function keys, so the billions of devices in the world with function keys now just don't get any benefit from them on the web anymore, full stop?

That's got to bother not only me, right? Please tell me I'm not the only one who is frustrated by that constraint? Why does a single vendor's design decision get to limit the affordances for an audience with already limited affordances. I truly reject this constraint as being a defacto constraint, but for some reason a lot of folks seem ok with it, which puzzles me.

If you have function keys on your platform, then use them, and if you don't, then map to something more appropriate. Is the lack of availability of home/end without modifiers on a Mac next? Will it simply be wrong to support home and end moving to beginnings or ends of lists? This is simply a race to the bottom where we are letting competing design decisions amongst vendors slowly carve away anything other than the most simplistic affordances in the name of cross-platform compatibility. That feels deeply wrong to me and I reject it.

Perhaps off topic for this issue, but also not off topic because of your keystroke request.

**Comment by StommePoes** [comment]
I was away :) I'm personally a ctrl-L person for reaching the address bar, but F6 allows me access to the OTHER thingies in the browser chrome, like plugins. 

I was working with a team at $previousEmployer where we were trying for F6 (or more likely, ctrl+F6) for a multi-pane dashboardy thing for teachers. Each "pane" wasn't necessarily a landmark, but each section had either a heading or a hidden heading; almost everything was some kind of form control or group of form controls, for creating content, pop quizzes and whatnot. This meant each area had just a crap-ton of focusables, so moving quickly to whole areas was necessary, and I had picked F6 because of its desktop history and somewhere in some guideline there is a single, brief mention :)

**Comment by carmacleod** [comment]
@sinabahram You are not alone in your frustrations about keyboard shortcuts. I tried not to be frustrated when I was writing down all the constraints for the "ideal" shortcut key, but what I was thinking is that there are no "ideal" shortcut keys available any more. 😢 

> As for not using function keys because of Mac, I'm really troubled by this.

No worries - I was just saying "ideally". In reality, we still need to use function keys. My Mac has physical function keys, so I have no experience with the new Mac Touch Bar... so I googled how VoiceOver works with it, and function keys are at least doable. The Touch Bar works kind-of like an iPhone - the user can hold down the Fn key (bottom left of keyboard) and either swipe or slide over the touch bar to select which function key to type, then double-tap to type it. It's sub-optimal (particularly while trying to hold down shift and command), but still possible.

I looked into some modifier+backtick/accent_grave key shortcuts, and they seem to be taken (in Chrome DevTools for go to console, and as a reserved shortcut on Mac... sigh - that's another type of constraint I forgot about earlier). Anyhow, I'm pretty sure that the only keys that even remotely make sense for landmarks are Ctrl/Cmd+F6 and Shift+Ctrl/Cmd+F6... which is why I am jealously protecting those keys. ;)

> I'm deeply frustrated by a similar notion that came from our treeview conversation that we must use the same keys across platforms. That one regarded enter and space.

The initial motivation for that was https://github.com/w3c/aria-practices/issues/1438, but we should keep that discussion over in https://github.com/w3c/aria/issues/1311. :)

> we were targeting Chrome rather heavily for that project, just due to Chromebooks and other reasons, ...

That makes sense. I got in trouble years ago for using Ctrl+L (Cmd+L on Mac) as the shortcut for "Go to Line" in our online code editor (I had reasoned that there were other ways to get there, like F6 and Alt+D... but apparently Cmd+L is the only way in Safari). A user made a scathing remark about us stealing the URL bar shortcut, so I guess I'm a bit sensitive about it. :)

@StommePoes 
> Each "pane" wasn't necessarily a landmark, but each section had either a heading or a hidden heading; almost everything was some kind of form control or group of form controls...

But they could've been landmarks, right? Given that a section with a heading (and an aria-labelledby pointing to the heading) is a landmark, and a form with an accessible name is a landmark, could all of those panes have been marked up as landmarks? Even the groups could have been wrapped in a div with role region and an aria-label?

@captainKAIO 
Sorry about the tangent! I'll try to circle back to the current issue about window splitter and F6...

In [comment 68 in the FF landmarks issue](https://bugzilla.mozilla.org/show_bug.cgi?id=670928#c68), Marco Z said:
> F6 already is being used in Firefox to switch between several major browser components, so cannot be used to also switch between landmarks on a page, or this will become very confusing.

So I'm guessing that he would say the same thing about using F6 for navigating between panes.

However, if you're up for it, it might be interesting to at least know whether F6 could somehow be integrated into the browser's "F6 ring" so that it would navigate to panes AND into (and out of) the browser trim (url bar, sidebar(s), tablist, bookmarks, etc). I'm only guessing that this might be doable by moving to the next/previous pane in an F6 key event handler and then calling event.preventDefault() _unless_ you were in the last pane on the page moving forward, or the first pane on the page moving backward.

I haven't tried this, and I'm sure there's something I haven't thought of, but if it could be made to work, maybe it would be an acceptable compromise because it's technically not stealing F6... just augmenting it with a few panes. :)

**Comment by StommePoes** [comment]
>But they could've been landmarks, right?

Only if it's okay to have like 20 landmarks on a page :)

But complex rocket-driving dashboards aren't web pages so some of these rules we tell each other don't land neatly (another complex dashboard I ran into was in an audit and yes it was used by, among others, NASA)

**Comment by carmacleod** [comment]
> Only if it's okay to have like 20 landmarks on a page :)

Heh, good point, but might be better to have 20 landmark hops than 100 tabs. ;)

> some of these rules we tell each other don't land neatly

Agreed. The [2014 WebAIM survey](https://webaim.org/projects/screenreadersurvey5/#numlandmarks) indicated that somewhere between 4-6 landmarks was "optimal", although some folks said more than 10.

But by the time you have a header, footer, main, a couple of navs and an aside, you're already at 6... so if that was a hard limit, many pages would have no room left for anything app-specific in custom regions. Those app-specific regions might ultimately be the most useful ones of all with regard to jumping back and forth between all the dashboardy bits and panes that someone is actually using to get real work done.

I do wonder if keeping "pane navigation" separate from "landmark navigation" would be useful, or just complicate things. I don't know the answer... just asking the question out loud. :)

**Comment by captainKAIO** [comment]
I think that [HTML5 element and WAI-ARIA landmark roles easily navigable in Firefox](https://github.com/w3c/aria-practices/issues/130) discussion is usefull for this issue. F6 seems to be used for "level 1" landmarks, Panes can be "level 1" landmarks or not. Must we use F6 key for panes?

**Comment by carmacleod** [comment]
@captainKAIO Did you mean to link to https://bugzilla.mozilla.org/show_bug.cgi?id=670928 ?
If so, which comment were you referring to?

**Comment by captainKAIO** [comment]
I'm refering to [comment 59](https://bugzilla.mozilla.org/show_bug.cgi?id=670928#c59) which present F6 and ctrl+F6 key on firefox.
if a pane has no focus, where F6 key goes ?

**Comment by mcking65** [comment]
I just raised issue #1546 with a proposal for addressing the key assignment issue aspect of the window splitter pattern as well as several other similar patterns.

**Comment by ZoeBijl** [comment]
@mcking65 I asked around and it turns out that <kbd>Esc</kbd> is supposed to bring focus back to the main content area if you’re in the browser chrome on macOS. This is implemented in Safari but not Chrome. Which makes me think that’s a Chrome bug.

Steps to reproduce:

1. Open Safari
2. Navigate to any page
3. Use the <kbd>Tab</kbd> key to navigate to some random element
4. Press <kbd>Command+L</kbd> to go to the address bar
5. Press <kbd>Esc</kbd> to close the suggestion panel
6. Press <kbd>Esc</kbd> to return focus to the element that last had focus on the page

**Comment by captainKAIO** [comment]
I have changed my html code to have landmark panes. I implemented also fixed panes on my example. 
Note : To toggle with the mouse or touch on fixed pane, I chose to apply the following effect: if the movement exceeds half of the pane then toggle

**Comment by patrickhlauke** [comment]
Randomly, I came across a test implementation that seemed to follow the pattern described in https://www.w3.org/WAI/ARIA/apg/patterns/windowsplitter/ ... but unfortunately it didn't work well at all in Chrome/Firefox and Jaws/NVDA (and the various combinations thereof). with `tabindex="0"` on the splitter, to actually make it focusable, neither AT went into forms mode by themselves, so had to be manually switched into it to make the cursor keys work. Jaws outright refused to announce the current value, and did not announce any change to it. So real-world support for this seems a bit flakey at best at the moment.

**Comment by MrGrigri** [comment]
> I've got a working drag example on [CodePen](https://codepen.io/MrGrigri/pen/vvEmYP). I was working on the drag part first and then I was going to add the keyboard interaction. I'm using CSS variables to control the width of the sidebar.

I have updated [my example](https://codepen.io/MrGrigri/pen/vvEmYP) to be fully keyboard compliant. I believe that a couple of things should be added to the specification. I have added an `aria-orientation` to the separator as mine will dynamically shift between vertical and horizontal depending on the viewport's width. The other thing could be some sort of step value. I move the separator every five pixels, but this is customizable. And from what I have researched there isn't an `aria-stepvalue` attribute to indicate to the user the amount that each arrow key would nudge the separator.

**Comment by patrickhlauke** [comment]
> Randomly, I came across a test implementation that seemed to follow the pattern described in https://www.w3.org/WAI/ARIA/apg/patterns/windowsplitter/ ... but unfortunately it didn't work well at all in Chrome/Firefox and Jaws/NVDA (and the various combinations thereof). with `tabindex="0"` on the splitter, to actually make it focusable, neither AT went into forms mode by themselves, so had to be manually switched into it to make the cursor keys work. Jaws outright refused to announce the current value, and did not announce any change to it. So real-world support for this seems a bit flakey at best at the moment.

Incidentally, we ended up ditching `role="separator"` and instead turned it into a conceptual `role="slider"` (with an appropriately descriptive `aria-label` to convey it's a splitter/separator, and changing the `aria-orientation` to be perpendicular to the direction of the separator ... although not all SRs actually announce the orientation anyway, so might as well omit that and double-up keyboard behaviour so that left and down decrement, and up and right increment, regardless of actual visual orientation). this then finally got Jaws to actually announce the current value and any changes, and it makes Jaws and NVDA go into forms mode when focused.

Long story short: the currently documented pattern here has really flaky/broken real-world support. This should be noted in the pattern, and ideally be fed back to SRs.

**Comment by JAWS-test** [comment]
Related: https://github.com/FreedomScientific/standards-support/issues/237 and https://github.com/FreedomScientific/standards-support/issues/489

**Comment by george-cz** [comment]
Hello @patrickhlauke, I've been recently building a resizable panel library for a major Microsoft product and we ran into the same issue. I was initially using the `role="separator"` on the handle, however a quick investigation showed that the support for this role among VO, NVDA and Jaws is sub-par compared to the `"slider"` to the point where it's unusable.

For now we've gone with the `"slider"` role, however I want to follow up on your note "the currently documented pattern here has really flaky/broken real-world support. This should be noted in the pattern, and ideally be fed back to SRs.".

Are you aware of any work in progres for any of the mentioned screen readers to fix the support for `"separator"`? Have you followed up with the SRs on any capacity? 

Thanks!

**Comment by thebabydino** [comment]
Is this issue still alive/ anyone interested in guiding me?

Last year, I saw someone ask [a question on reddit](https://www.reddit.com/r/css/comments/1ohqiu8/simple_3_panel_layout/) on how to create something with separators for a presentation. While the question was not about accessibility, it ended up sending me down a rabbit hole.

I eventually dropped it at some point as I was not happy with how it worked with screen readers - you can see my summary of my screen reader testing (Ubuntu only, I do not own any device new enough to properly work with anything non-Linux) on the details page of the CodePen demo https://codepen.io/thebabydino/details/gbPBxJP

Now I happened to see there's an APG pattern for this at the very bottom of the list of patterns and was curious to check out the code there to see how I could improve mine because I still had an itch to scratch with this demo... and discovered there was no demo for this pattern.

Here are a few things that confuse me after seeing the guidelines on that page and reading this discussion. Maybe someone could clarify these things to let me know how I should update my demo.

*(which also needs some work on the JS because currently the JS is based on some assumptions that won't be true for every separator case out there... but that part I can just deal with myself once my confusion over some basics gets cleared)*

## The "primary pane" concept

My simplified structure is following the one from the reddit question about a slideshow.

![illustration from reddit question, described below](https://github.com/user-attachments/assets/abb1a704-6e44-4c22-8203-a1ace3c84771)

On the top left, there's an index/ a table of contents going through the slides. On the top right, there is the current slide. There is a draggable separator between them. At the bottom, there are the slide notes for the current slide. There is a draggable separator between the slide notes and all else above.

The simplified structure would look like this:

```
<nav id='navi'></nav>
<div role='separator' tabindex='0' 
	aria-valuemin='0' aria-valuenow='50' aria-valuemax='100' 
	aria-controls='navi' aria-orientation='vertical' 
	aria-label='separator between navigation and main content'></div>
<main id='main'></main>
<div role='separator' tabindex='0' 
	 aria-valuemin='0' aria-valuenow='50' aria-valuemax='100' 
	 aria-controls='side' aria-orientation='horizontal' 
	 aria-label='separator between everything and notes'></div>
<aside id='side'></aside>
```

The `nav` holds the index/ table of contents.

The `main` holds the current slide.

The  `aside` holds the current slide notes.

The vertical separator (the first one in the markup) is the vertical bar between the `nav` (index/ table of contents) panel and the `main` (current slide) panel.

The horizontal separator (the first one in the markup) is the horizontal bar between the top panels - the `nav` (index/ table of contents) panel and the `main` (current slide) panel - and the bottom `aside` (current slide notes) panel.

Initially, when I read about "primary panel", I first thought that would be the `main` for both separators, but then I read this part:

> For example, consider a book reading application with a primary pane for the table of contents and a secondary pane that displays content from a section of the book. The two panes are divided by a vertical splitter labelled "Table of Contents". When the table of contents pane has its maximum size, the splitter has a value of 100, and when the table of contents is completely collapsed, the splitter has a value of 0.
> 
> Note that the term "primary pane" does not describe the importance or purpose of content inside the pane.

This would mean the primary panel for the vertical separator is the `nav` and the primary panel for the horizontal separator is the `aside`? This is what my current HTML assumes and it would be great if someone could confirm it.

## The vertical vs. horizontal separator concept

Currently, I have `aria-orientation` set to `vertical` for the vertical separator line and to `horizontal` for the horizontal separator line.

However, the vertical separator line moves along the *x* (horizontal) axis and the horizontal separator line moves along the *y* (vertical) axis, so I'm not sure whether I got the orientation values right or they should be the other way around.

I would think I got it right based on the keyboard interaction guide:

> * `Left Arrow`: Moves a vertical splitter to the left.
> * `Right Arrow`: Moves a vertical splitter to the right.
> * `Up Arrow`: Moves a horizontal splitter up.
> * `Down Arrow`: Moves a horizontal splitter down.

... but then I think about it again and I'm confused again.

## The "collapsed" pane

In the keyboard interaction section, there's this part:

> * `Enter`: If the primary pane is not collapsed, collapses the pane. If the pane is collapsed, restores the splitter to its previous position.

What does "collapsed" mean in this context?

Zero the size of the pane along the axis of separator motion? What if that's not allowed by the design constraints?

Or is it the minimum size (the one at `aria-valuemin='0'`)? In this case, the effect of `Enter` if the pane is not collapsed would be the same as the effect I have implemented for `Home`, right?

## "Cycling" though the panes

There's also this part:

> * `F6`: (Optional): Cycle through window panes.

What does cycling mean in this scenario? What should happen when cycling through window panes? Some of the participants in this thread have mentioned using this, what does/ should it do?

Asking this as I am not a Slack user/ keyboard user in general. I'm a mouse/ trackpad user because I could never really understand keyboard use beyond typing in a text box. I know about things like using the arrow keys to increment decrement values for sliders and things and going from link to link with Tab, though I don't even really do that myself outside of testing my code because it just feels inconvenient and slow as opposed to using the cursor.

But `F6`... I have no idea what it does. I've tried pressing  it on various pages and I haven't seen anything happen, though tbh, I have no idea how to even check if any of the F keys do anything in any scenario.

---

I guess these are my questions for now. It would be great if someone could guide me in the right direction, then I could try to produce a working demo.

To anyone who could find the time to read this and answer my questions: thank you!

**Comment by JAWS-test** [comment]
@thebabydino 

> This would mean the primary panel for the vertical separator is the nav and the primary panel for the horizontal separator is the aside? This is what my current HTML assumes and it would be great if someone could confirm it.

From what I understand from the explanation on the page https://www.w3.org/WAI/ARIA/apg/patterns/windowsplitter/, you can decide for yourself which window is the primary window. But once you’ve chosen a primary window, the three ARIA attributes `aria-valuenow`, `aria-valuemin`, and `aria-valuemax` refer to that window. That’s why your example is incorrect: in your case, `aria-valuemin` and `aria-valuemax` are 0 and 100, but I can’t maximize an area to full screen or hide it completely in your example. In your example, for the first splitter, `aria-valuemin` should be around 10 and `aria-valuemax` around 40 (you’ll have to determine the exact numbers yourself; I’ve only estimated them here). `aria-valuenow` should then only be allowed to range between 10 and 40.

**Comment by JAWS-test** [comment]
@thebabydino 

> However, the vertical separator line moves along the x (horizontal) axis and the horizontal separator line moves along the y (vertical) axis, so I'm not sure whether I got the orientation values right or they should be the other way around.

The `aria-orientation` attribute does not specify the direction of arrow key navigation, but rather the orientation of the element (https://www.w3.org/TR/wai-aria-1.3/#aria-orientation). Therefore, the vertical splitter should also be marked up as vertical.

**Comment by JAWS-test** [comment]
@thebabydino 
> What does "collapsed" mean in this context?

That means: completely hidden, invisible. That’s not possible in your example. And that’s okay. Not every splitter allows movement between 0 and 100. In your case, for example, it’s between 20 and 40. In that sense, ENTER should only work if complete hiding is possible. But I think the ENTER thing is nonsense; it should be an optional feature at most. For users, it’s much more intuitive and familiar to reach the maximum and minimum values using HOME and END—and that’s already implemented correctly in your example. 

It would be more important to have a different method for efficiently operating the splitter, which has not yet been mentioned on https://www.w3.org/WAI/ARIA/apg/patterns/windowsplitter/. Using PAGE UP and PAGE DOWN should make it possible to make larger jumps with the splitter (e.g., 10% of the total range of motion). Otherwise, I might have to press the arrow key 100 times to set the desired size.

**Comment by JAWS-test** [comment]
@thebabydino 

> What does cycling mean in this scenario? 

In desktop applications, I can use F6 to navigate between different sections (https://learn.microsoft.com/en-us/windows/apps/design/accessibility/keyboard-accessibility?#keyboard-navigation-between-application-panes-with-f6). This concept has been adopted by web applications. So, in your example, pressing F6 should take you to the next landmark region, and SHIFT+F6 to the previous one. This is important for enabling efficient navigation between sections for keyboard users (see: https://github.com/w3c/aria-practices/issues/1546).

Note: Since F6 is used by the browser itself (which is a desktop application), some web applications use F6 (thereby “stealing” the key from the browser) or use CTRL+F6 instead, so that F6 remains available for the browser (example: https://support.microsoft.com/en-us/office/use-outlook-web-app-email-with-a-screen-reader-d039e28a-3238-4214-bdb0-3a2533a7d7a2, chapter: "Landmarks and navigation"). 

The ARIA APG page should, in its recommendations, opt for one of the two methods whenever possible or describe both methods as permissible, but should not recommend F6 for one element and CTRL+F6 for another, as that would be confusing.

**Comment by JAWS-test** [comment]
@thebabydino 

> ... and discovered there was no demo for this pattern.

I would really appreciate seeing a working example on the ARIA APG page soon, and I don’t understand why one isn’t available yet. It is, after all, a very simple control. However, before you invest your time, please wait for approval from the ARIA Working Group (I am not part of it and therefore cannot say whether your example is good enough to be adopted by ARIA APG). 

I would think it would be good if such an example also included content (e.g., a navigation bar on the left, a table with many records on the right, and a detailed view of a selected record at the bottom)—it doesn’t have to be interactive, but it would then demonstrate a typical use case.
